### DISCUSSION TITLE
Requirements for international number formatting

### DISCUSSION THREAD

**Discussion by cwegener** [apache/superset#23684] [Ideas]
**Problem**

Superset currently uses `d3-format` default settings for international number formatting. I.e. every number that is getting formatted with one of the `d3-format` options in Superset, uses the **English US** (`en-US` in BCP47 notation) settings for:

- the decimal point (e.g., ".").
- the group separator (e.g., ",").
- the array of group sizes (e.g., [3]), cycled as needed.
- the currency prefix and suffix (e.g., ["$", ""]).

This prevents the greater adoption of Superset by users who have different expectations for number formatting.

The main problem exists for users in regions who use the exact opposite configuration of the *decimal point* and the *group separator*, i.e. all German speaking regions use `.` as the group separator and `,` as the decimal point, which will cause confusion for those user groups in the current configuration.

**Solutions**

There is one open PR that attempted one possible solution. #20075

That PR made 2 major design decisions, which should be reviewed before implementing such feature. These 2 design decisions are:

1. Forgo the use of commonly present language tags (as seen in IETF BCP47, i.e. `de`, `de-AU`, `de-CH` or `de-DE` in your case) in favor of manually creating custom D3 number formatting rules via a new `D3_FORMAT` Flask configuration property.
2. Have exactly one number formatting setting for the whole Superset instance. All dashboards and charts will use the same number formatting setting. Considerations about different number formatting settings on a per dashboard or per chart basis should be considered either for an initial PR or as a future enhancement.

What are everyone's thoughts about the above 2 decisions?

(I am assuming that the overall feature of changing international number formatting options in general is something that we can all agree on is needed :grin:)

**Comment by lf-floriandin** [comment]
I strongly agree with 1.
Using language tags would be way more comfortable and easier to use from a user perspective.
I also agree with 2.
Initially a global setting would be enough, but longterm it would be nice to change it on a per dashbord or even per chart basis.

**Comment by sfirke** [comment]
1 sounds good to me.  On a different FOSS project facing this challenge, we ended up going with user-specification rather than trying to pick up via locale or language.

If there's an opportunity to address other formatting pain points like https://github.com/apache/superset/issues/16954, https://github.com/apache/superset/issues/16610, https://github.com/apache/superset/issues/7470 during this process - or at least pave the way for the future - that would be swell.

**Comment by sfirke** [comment]
I believe this was implemented in https://github.com/apache/superset/pull/20075 which appeared in Superset version 3.0.0.  People following this discussion should see that pull request discussion for examples of how to implement this in their deployment of Superset.

**Comment by Hyperrick** [comment]
> **Problem**
> 
> Superset currently uses `d3-format` default settings for international number formatting. I.e. every number that is getting formatted with one of the `d3-format` options in Superset, uses the **English US** (`en-US` in BCP47 notation) settings for:
> 
> * the decimal point (e.g., ".").
> * the group separator (e.g., ",").
> * the array of group sizes (e.g., [3]), cycled as needed.
> * the currency prefix and suffix (e.g., ["$", ""]).
> 
> This prevents the greater adoption of Superset by users who have different expectations for number formatting.
> 
> The main problem exists for users in regions who use the exact opposite configuration of the _decimal point_ and the _group separator_, i.e. all German speaking regions use `.` as the group separator and `,` as the decimal point, which will cause confusion for those user groups in the current configuration.
> 
> **Solutions**
> 
> There is one open PR that attempted one possible solution. #20075
> 
> That PR made 2 major design decisions, which should be reviewed before implementing such feature. These 2 design decisions are:
> 
> 1. Forgo the use of commonly present language tags (as seen in IETF BCP47, i.e. `de`, `de-AU`, `de-CH` or `de-DE` in your case) in favor of manually creating custom D3 number formatting rules via a new `D3_FORMAT` Flask configuration property.
> 2. Have exactly one number formatting setting for the whole Superset instance. All dashboards and charts will use the same number formatting setting. Considerations about different number formatting settings on a per dashboard or per chart basis should be considered either for an initial PR or as a future enhancement.
> 
> What are everyone's thoughts about the above 2 decisions?
> 
> (I am assuming that the overall feature of changing international number formatting options in general is something that we can all agree on is needed 😁)

Hi, have you found a way to change the number formatting based on the current user's language preference? In my use case, I have users with different language preferences, such as English, German, etc., which would require different formatting of the charts in the frontend when it comes to decimal separators or thousands separators. I haven't found a way to link the formatting to a locale variable yet.
